petastorm/reader_impl/reader_v2.py

```diff
@@ -0,0 +1,320 @@
+                       pass in a unique partition number in the range [0, num_training_partitions).
+                       num_training_partitions must be supplied as well.
+        :param num_training_partitions An int denoting the number of training partitions (how many nodes are performing
+                       the multi node training)
+        :param read_timeout_s: A numeric with the amount of time in seconds you would like to give a read before it
+                       times out and raises an EmptyResultError. Pass in None for an infinite timeout
+        :param cache: An object conforming to `cache.CacheBase` interface. Before loading row groups from a parquet file
+                       the Reader will attempt to load these values from cache. Caching is useful when communication
+                       to the main data store is either slow or expensive and the local machine has large enough storage
+                       to store entire dataset (or a partition of a dataset if num_training_partitions is used).
+        :param decoder_pool: An instance of a concurrent.futures pool executor used for decoding. If None,
+          a default ThreadPoolExecutor(5) will be used.
+        :param loader_pool: An instance of a concurrent.futures pool executor used for decoding. If None,
+          a default ThreadPoolExecutor(5) will be used.
+        :param shuffle_options : ShuffleOptions object to describe how to shuffle dataset (supercedes shuffle parameter)
+                       defaults to shuffling row groups but not to drop rows based on partitions.
+        :param shuffle: DEPRECATED boolean whether to shuffle the row group order. Use shuffle_row_groups in
+                       ShuffleOptions instead.
+
+        By default, `NullCache` implementation
+        """
+
+        # 1. Resolve dataset path (hdfs://, file://) and open the parquet storage (dataset)
+        # 2. Get a list of all groups
+        # 3. Filter rowgroups
+        #    a. predicates
+        #    b. row-group selector (our indexing mechanism)
+        #    c. partition: used to get a subset of data for distributed training
+        # 4. Launch a new thread running `worker_loop` function.
+
+        if dataset_url is None or not isinstance(dataset_url, six.string_types):
+            raise ValueError("""dataset_url must be a string""")
+
+        if not (isinstance(schema_fields, collections.Iterable) or isinstance(schema_fields, NGram)
+                or schema_fields is None):
+            raise ValueError("""Fields must be either None, an iterable collection of Unischema fields or an NGram
+            object.""")
+
+        if sequence is not None:
+            raise ValueError("""'sequence' argument of Reader object is deprecated. Please pass an NGram instance to
+            'schema_fields' argument instead.""")
+
+        self.ngram = schema_fields if isinstance(schema_fields, NGram) else None
```

Comment:
This is missing some validation for overlap https://github.com/uber/petastorm/blob/master/petastorm/reader.py#L120